ngraph/core/reference/include/ngraph/runtime/reference/cum_sum.hpp

```diff
@@ -5,107 +5,41 @@
 #pragma once
 
 #include <cmath>
-#include <map>
-#include <utility>
-#include <vector>
-
-#include "ngraph/coordinate_transform.hpp"
-#include "ngraph/type/bfloat16.hpp"
-#include "ngraph/type/float16.hpp"
 
 namespace ngraph {
 namespace runtime {
 namespace reference {
+
 template <typename T, typename P>
 void cumsum(const T* arg,
             const P* axis_tensor,
             T* out,
             const Shape& tensor_shape,
             const bool exclusive,
             const bool reverse) {
-    NGRAPH_SUPPRESS_DEPRECATED_START
-    CoordinateTransform temp_transform(tensor_shape);
-    for (const Coordinate& output_coord : temp_transform) {
-        out[temp_transform.index(output_coord)] = 0;
-    }
-
-    P axis = axis_tensor[0];
-    P rank = tensor_shape.size();
-
-    if (axis < -rank || axis > rank) {
-        throw ngraph_error("axis must be in the range [-rank, rank]");
-    }
-    axis = axis < 0 ? rank + axis : axis;
-
-    auto get_key = [&, axis](const Coordinate& coord) -> Coordinate {
-        Coordinate result(coord.size(), 0);
-        result[axis] = coord[axis];
-
-        for (size_t i = 0; i < coord.size(); i++) {
-            result[i] = coord[i] - result[i];
-        }
-        return result;
-    };
-
-    auto update_output_buffer =
-        [&](size_t input_index, size_t output_index, T& prev, std::vector<std::pair<size_t, T>>& tensor_vec) -> void {
-        tensor_vec[input_index].second = prev + tensor_vec[input_index].second;
-        out[tensor_vec[output_index].first] = tensor_vec[input_index].second;
-
-        // update prev to hold the last result value to compute ruuning sum for
-        // subsequent iter
-        prev = out[tensor_vec[output_index].first];
-    };
-
-    auto cum_sum = [&, exclusive, reverse](std::vector<std::pair<size_t, T>>& tensor_vec) {
-        if (!reverse) {
-            T prev = 0;
-            for (size_t i = 0; i < tensor_vec.size(); i++) {
-                if (exclusive && i == 0) {
-                    out[tensor_vec[i].first] = prev;
-                    continue;
-                }
-                // we will compute running sum of j-1 elements if exlusive=1 or else
-                // for j elements if exclusive = 0
-                size_t arg_index = exclusive == 1 ? i - 1 : i;
-                update_output_buffer(arg_index, i, prev, tensor_vec);
+    const auto rank = tensor_shape.size();
+    const auto axis = axis_tensor[0] >= 0 ? axis_tensor[0] : rank + axis_tensor[0];
+    const auto axis_dim = tensor_shape[axis];
+
+    const auto size_before_axis = shape_size(Shape(tensor_shape.begin(), tensor_shape.begin() + axis));
+    const auto size_after_axis = shape_size(Shape(tensor_shape.begin() + axis + 1, tensor_shape.end()));
+
+    const auto reverse_shift = reverse ? -1 : 1;
+    const auto element_shift = exclusive ? size_after_axis * reverse_shift : 0;
+
+    for (auto i = 0; i < size_before_axis; ++i) {
```

Comment:
Since `size_before_axis` is `size_t`, I suggest using this type for `i` explicitly (the same applies to the nested loops below.

Reply:
Thanks, done